analysis/src/github.rs

```diff
@@ -0,0 +1,172 @@
+//! This module abstracts the communication with GitHub API for a given crate
+
+use anyhow::{anyhow, Result};
+use guppy::graph::PackageMetadata;
+use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, USER_AGENT};
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+use url::Url;
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct RepoStats {
+    pub full_name: Option<String>,
+    pub stargazers_count: u64,
+    pub subscribers_count: u64,
+    pub forks: u64,
+}
+
+pub struct GitHubReport {
+    pub name: String,       // name of the crate
+    pub repository: String, // repository url
+    pub is_github_repo: bool,
+    pub repo_stats: RepoStats,
+}
+
+impl GitHubReport {
+    fn new(name: String, repository: String) -> Self {
+        //Returns a default GitHubReport with is_github_repo set as false
+        GitHubReport {
+            name,
+            repository,
+            is_github_repo: false,
+            repo_stats: RepoStats {
+                full_name: None,
+                ..Default::default()
+            },
+        }
+    }
+}
+
+pub struct GitHubAnalyzer {
+    client: reqwest::blocking::Client,
+}
+
+impl GitHubAnalyzer {
+    fn construct_headers() -> Result<HeaderMap> {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static("diem/whackadep"));
+
+        let pat = std::env::var("GITHUB_TOKEN")?;
+        let pat = format!("token {}", pat);
+        let mut auth_value = HeaderValue::from_str(&pat)?;
+        auth_value.set_sensitive(true);
+        headers.insert(AUTHORIZATION, auth_value);
+
+        Ok(headers)
+    }
+
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            client: reqwest::blocking::Client::builder()
+                .default_headers(Self::construct_headers()?)
+                .build()?,
+        })
+    }
+
+    pub fn analyze_github(self, package: &PackageMetadata) -> Result<GitHubReport> {
+        let name = package.name();
+        let repository = package
+            .repository()
+            .ok_or_else(|| anyhow!("no repository present for crate {}", name))?;
+
+        let is_github_repo = Self::is_github_url(&repository);
+        if !is_github_repo {
+            return Ok(GitHubReport::new(name.to_string(), repository.to_string()));
+        }
+
+        let repo_fullname = Self::get_fullname_from_repository_url(repository)?;
```

Comment:
Lets combined these steps (and the functions below). Right now you're checking if the url host is "github.com" twice. We should simplify that such that we're only doing the check once.

Reply:
let github_repo_name = if let Ok(name) = get_github_repo_name(..) {
    name
} else {
    // return empty report
}